h2/src/main/org/h2/mvstore/tx/TransactionStore.java

```diff
@@ -476,31 +565,49 @@ void commit(Transaction t, long maxLogId, int oldStatus) {
      * End this transaction
      *
      * @param t the transaction
-     * @param oldStatus status of this transaction
+     * @param hasChanges false for R/O tx
      */
-    synchronized void endTransaction(Transaction t, int oldStatus) {
-        if (oldStatus == Transaction.STATUS_PREPARED) {
-            preparedTransactions.remove(t.getId());
-        }
+    synchronized void endTransaction(Transaction t, boolean hasChanges) {
+        int txId = t.transactionId;
         t.setStatus(Transaction.STATUS_CLOSED);
-        openTransactions.clear(t.transactionId);
-        if (oldStatus == Transaction.STATUS_PREPARED || store.getAutoCommitDelay() == 0) {
-            store.tryCommit();
-            return;
-        }
-        // to avoid having to store the transaction log,
-        // if there is no open transaction,
-        // and if there have been many changes, store them now
-        if (undoLog.isEmpty()) {
-            int unsaved = store.getUnsavedMemory();
-            int max = store.getAutoCommitMemory();
-            // save at 3/4 capacity
-            if (unsaved * 4 > max * 3) {
+
+        boolean success = transactions.compareAndSet(txId, t, null);
+        assert success;
```

Comment:
The same strange assignment that is not used below.

Reply:
This assignment cannot be inlined, but it also takes a some time to check that `success` is overwritten before read below. If stop condition in the following loop will be inlined the variable here should be OK.

Reply:
Same case as in my other comment above